### ISSUE TITLE
`findOne` with a scope option

### ISSUE THREAD

**Issue by morishjs** [iWinston/typeorm-plus#7]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Hi. I tried to get a deleted entity from the database. But it seems that `findOne` with scope doesn't work. 

```javascript
const deleteId = product.id;
// some codes for deleting the product...
// ...

const deletedProduct = await getRepository(Product).findOne(deleteId, {scope: 'deleted'});
expect(deletedProduct).toBeNull(); // because `deletedProduct` is undefined.
```

The query shows that the scope option is not properly applied to the query builder. (`AND Product.deletedAt IS NULL`)
```SQL
    SELECT DISTINCT `distinctAlias`.`Product_id` as "ids_Product_id" FROM (SELECT `Product`.`id` AS `Product_id`, `Product`.`name` AS `Product_name`, `Product`.`description` AS `Product_description`, `Product`.`price` AS `Product_price`, `Product`.`detailLink` AS `Product_detailLink`, `Product`.`brandName` AS `Product_brandName`, `Product`.`imageLinks` AS `Product_imageLinks`, `Product`.`createdAt` AS `Product_createdAt`, `Product`.`updatedAt` AS `Product_updatedAt`, `Product`.`deletedAt` AS `Product_deletedAt`, `Product_brand`.`name` AS `Product_brand_name`, `Product_brand`.`url` AS `Product_brand_url`, `Product_brand`.`createdAt` AS `Product_brand_createdAt`, `Product_hashtags`.`name` AS `Product_hashtags_name`, `Product_hashtags`.`createdAt` AS `Product_hashtags_createdAt` FROM `product` `Product` LEFT JOIN `brand` `Product_brand` ON `Product_brand`.`name`=`Product`.`brandName`  LEFT JOIN `product_hashtags_hashtag` `Product_Product_hashtags` ON `Product_Product_hashtags`.`productId`=`Product`.`id` LEFT JOIN `hashtag` `Product_hashtags` ON `Product_hashtags`.`name`=`Product_Product_hashtags`.`hashtagName` WHERE `Product`.`id` IN ("1993bf98-e902-41cc-989c-d871ca0adbf4") AND `Product`.`deletedAt` IS NULL) `distinctAlias` ORDER BY `Product_id` ASC LIMIT 1

```

Here is my entity class
```typescript
@Entity()
export class Product extends BaseEntity {
  public static scope = {
    deleted: {
      deletedAt: Not(IsNull()),
    },
  };

//some codes...
}
```

I don't know why the `find` works as expected but `findOne` doesn't.

**Comment by iWinston** [comment]
Sorry for the slow response. 
I was working hard to refactor all the code of TypeORM+ and add more tests for it.

Now the release 0.0.55 is available. And I have added the test for custom query scope using the `findOne`. You can check it if you want.

```
    it("should set custom query scope correctly", () => Promise.all(connections.map(async connection => {

        const post1 = new Post();
        post1.title = "title#1";
        const post2 = new Post();
        post2.title = "title#2";
        const post3 = new Post();
        post3.title = "title#3";

        await connection.manager.save(post1);
        await connection.manager.save(post2);
        await connection.manager.save(post3);

        const loadedWithCustomScopePosts = await connection
            .getRepository(Post)
            .find({
                scope: "title3",
            });

        loadedWithCustomScopePosts!.length.should.be.equal(1);
        loadedWithCustomScopePosts![0].title.should.be.equals("title#3");

        const loadedWithCustomScopePost = await connection
            .getRepository(Post)
            .findOne(3, {
                scope: "title3",
            });
        loadedWithCustomScopePost!.title.should.be.equals("title#3");

    })));
```

**Comment by morishjs** [comment]
@iWinston 
Thanks.
But it doesn't work in my code even though I already upgraded the latest release.
I will check what's happening under the hook on weekends. 

Could you hold this issue until I figure out what's the problem?

**Comment by iWinston** [comment]
Yes. of course.

And you can come back here if there's something I can help.

**Comment by morishjs** [comment]
Hi @iWinston 

I figured out what's happening in the package and why in my code `findOne` with the scope option doesn't work. 

https://github.com/iWinston/typeorm-plus/blob/043b9cce0e54b9053f655e24976cf04f089874ea/src/query-builder/SelectQueryBuilder.ts#L1905

That line `this.clone()` does not clone deeply so even the function `setScope` was called but `expressionMap`is reset after `this.clone()` is called.

I just put `setScope` to set the scope again in order to resolve the unexpected behavior.
But I don't know that's good to be clean code in the package.

By the way, I tried to submit the pull request, but I have no idea how to set up typeorm+ environment to run the test codes. Would you suggest the way to debug or run unit tests in your local?

**Comment by iWinston** [comment]
Thank you for figuring out this bug.

I think the best and simplest way to fix it is to  clone the `scope`  property when calling the `clone` method of `QueryExpressionMap.ts` like this:

```TypeScript
export class QueryExpressionMap {
     ...

    /**
     * Specifies what scope should be applied to the repository
     * If it is false, none of the scopes will be applied
     * If it is undefined, the value will be "default"
     */
    scope?: string | false;
    
    ...

    clone(): QueryExpressionMap {
        ...
        map.scope = this.scope;
        ...
    }
}
```

I run the tests using the script `npm run test`.

By the way,  can you show me the whole code reproducing the bug? So we can add more tests to cover this situation.

**Comment by morishjs** [comment]
There are nothing to be different with first code I've already mentioned in this issue.

Only way in which I can help you reproduce the error is 

1. `SelectQueryBuilder#getOne` is called
2. `getRawAndEntities#getRawAndEntities` is called
3. `executeEntitiesAndRawResults` is called
4. And... the below code makes the trouble.
```
rawResults = await this.clone()
                    .mergeExpressionMap({ extraAppendedAndWhereCondition: condition })
                    .setParameters(parameters)
                    .loadRawResults(queryRunner);
```

**Comment by iWinston** [comment]
Hi @morishjs 

I have reproduced the bug using `findeOne` with an eager relation.
Is there an eager relation in your entity?

**Comment by morishjs** [comment]
@iWinston 
> Is there an eager relation in your entity?

Yes!

**Comment by iWinston** [comment]
Hi @morishjs 

I have refactored the way to apply query scope to fix the bug that it can't set query scope with relation correctly. [Here is the commit](https://github.com/iWinston/typeorm-plus/commit/27c2b571010d86b60c4f79e3d5f3ed47ab80cd8b)

And I added some tests for this situation.


Now TypeORM+ 0.0.57 is available, so you can upgrade to 0.0.57 and fix the bug.

**Comment by morishjs** [comment]
@iWinston 
It works perfectly. Thank you for being patient with fixing the bug.
I am closing this issue.
